util/project_allocation_mapper.py

```diff
@@ -113,20 +113,21 @@ def _create_ticket_for_pi_request(self, user):
                                  requestor = "us@tacc.utexas.edu")
         rt.createTicket(ticket)
 
+    '''
+    get all projects
+    returns list of tas_proj_obj, sorted by most recent allocation
+    allocations in each object are sorted newest to oldest
+    
+    '''
     def get_all_projects(self):
-        projects = {}
-        project_pi_info = {}
-        for proj in portal_proj.objects.select_related('pi').all():
-            project_pi_info[proj.charge_code] = self.portal_user_to_tas_obj(proj.pi)
-        for alloc in portal_alloc.objects.select_related('project', 'project__pi', 'project__type', 'project__field', 'requestor', 'reviewer').all():
-            proj = self.portal_to_tas_proj_obj(alloc.project,
-                                               fetch_allocations=False,
-                                               fetch_balance=False,
-                                               pi_info=project_pi_info[alloc.project.charge_code])
-            if proj['chargeCode'] not in projects:
-                projects[proj['chargeCode']] = proj
-            projects[proj['chargeCode']]['allocations'].append(self.portal_to_tas_alloc_obj(alloc))
-        return sorted(list(projects.values()), reverse=True, key=self.sort_by_allocation_request_date)
+        #each project has many allocations; each allocation has a 'date_requested'
+        #for each project, get the most recent date among its allocations, then annotate with 'newest_request'
+        #sort list of projects by 'newest_request'
+        from django.db.models import Max
+        sorted_projects = portal_proj.objects.annotate(newest_request=Max('allocations__date_requested')).order_by('newest_request').reverse().select_related('pi')     
```

Comment:
You might be able to sneak in a `.prefetch_related('allocations')` to really speed this up. Not sure how Django will treat the annotation, which asks for the related allocations. There is some way to turn on logging such that you can see the precise SQL queries Django is issuing, which could prove the value. You could also throw it on and see if it likes it. This is used elsewhere to reduce extra queries in Trovi: https://github.com/ChameleonCloud/portal/blob/0384d1247fc09f712fda7b15de87e3be2de5d1fd/sharing_portal/views.py#L164-L175

`prefetch_related` is like `select_related` but it can handle ManyToMany. It _may_ work to just pass 'allocations' into the `select_related`, I don't know Django well enough to be sure.

Reply:
real performance improvements need more refactoring
1. no optimization ~7s
2. select_related('type','pi','field') 3.8-4.5s
3. prefetch_related('allocations) 4-4.5s
4. both: 4-4.5s

new view with no frontend, simple conversion from model ->dict
1. json.dumps(portal_proj.objects.all() ) 0.4s
2. json.dumps(portal_alloc.objects.all() ) 0.5s

Easiest win is just restricting the filterset by default, make a new view for:
"projects with at least one pending allocation"